### PR TITLE
TRACK-780 Add storage location management to admin UI

### PR DESCRIPTION
**Pull request by sgrimm**
Currently, the only way to configure the list of storage locations that show up
in the seed bank UI is to connect to the database and run SQL commands by hand.

Since we're in the middle of rethinking how we model facilities and such, it
doesn't make sense to add storage location management to the frontend. But as a
stopgap, we at least need a way to set things up for customers.

Add a section to the facility page in the admin UI where we can manage the list
of storage locations for a facility.

### CODE REVIEW

**Comment by sgrimm** [comment]
Current dependencies on/for this PR:
* main
  * **PR #194** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/194" target="_blank">Graphite</a>  (this PR)
    * **PR #195** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/195" target="_blank">Graphite</a> 
      * **PR #196** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/196" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

**Comment by sgrimm** [comment]
@kallen07 This is bigger than you might expect because I'm assuming at some point we're going to need to make the list of storage locations customer-editable, so I built some of the scaffolding we'll need (permission checks, etc.) to support adding API endpoints. That stuff all follows the usual patterns, so hopefully there aren't any surprises here.

File: src/main/kotlin/com/terraformation/backend/customer/api/AdminController.kt [hunk 1]
```diff
@@ -597,6 +603,52 @@ class AdminController(
...
+      @RequestParam("facilityId") facilityId: FacilityId,
+      @RequestParam("name") name: String,
+      @RequestParam("condition") condition: StorageCondition,
+      redirectAttributes: RedirectAttributes
+  ): String {
+    facilityStore.createStorageLocation(facilityId, name, condition)
+
+    redirectAttributes.addFlashAttribute("successMessage", "Storage location created.")
+
+    return facility(facilityId)
```
Comment (kallen07) [inline src/main/kotlin/com/terraformation/backend/customer/api/AdminController.kt:617]: How come we return the facility ID from each of these functions?

File: src/main/kotlin/com/terraformation/backend/customer/model/IndividualUser.kt [hunk 2]
```diff
@@ -387,6 +388,29 @@ data class IndividualUser(
 
   override fun canUpdateTimeseries(deviceId: DeviceId): Boolean = canCreateTimeseries(deviceId)
 
+  override fun canCreateStorageLocation(facilityId: FacilityId): Boolean {
+    return when (facilityRoles[facilityId]) {
+      Role.OWNER, Role.ADMIN -> true
```
Comment (kallen07) [inline src/main/kotlin/com/terraformation/backend/customer/model/IndividualUser.kt:393]: Are you basing these permissions based on previous patterns, or is there documentation about storage location permissions?

File: src/main/kotlin/com/terraformation/backend/customer/model/IndividualUser.kt [hunk 3]
```diff
@@ -387,6 +388,29 @@ data class IndividualUser(
...
+      else -> false
+    }
+  }
+
+  override fun canReadStorageLocation(storageLocationId: StorageLocationId): Boolean {
+    val facilityId = parentStore.getFacilityId(storageLocationId) ?: return false
+    return facilityId in facilityRoles
+  }
+
+  override fun canUpdateStorageLocation(storageLocationId: StorageLocationId): Boolean {
```
Comment (kallen07) [inline src/main/kotlin/com/terraformation/backend/customer/model/IndividualUser.kt:403]: What do you think about making `canUpdateStorageLocation` call `canCreateStorageLocation`?

File: src/main/kotlin/com/terraformation/backend/customer/db/FacilityStore.kt [hunk 4]
```diff
@@ -128,4 +134,70 @@ class FacilityStore(
...
+          .set(NAME, name)
+          .execute()
+    }
+  }
+
+  /**
+   * Deletes a storage location. This will only succeed if the storage location is not referenced by
+   * any accessions.
+   *
+   * @throws org.springframework.dao.DataIntegrityViolationException The storage location is in use.
```
Comment (kallen07) [inline src/main/kotlin/com/terraformation/backend/customer/db/FacilityStore.kt:193]: Is it being enforced at the database level that you cannot delete storage location if there's an accession references it?

File: src/test/kotlin/com/terraformation/backend/customer/db/FacilityStoreTest.kt [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            modifiedBy = user.userId,
+            modifiedTime = newTime,
+            name = "New Name",
+        )
+
+    val actual = storageLocationsDao.fetchOneById(storageLocationId)
+
+    assertEquals(expected, actual)
+  }
+}
```
Comment (kallen07) [inline src/test/kotlin/com/terraformation/backend/customer/db/FacilityStoreTest.kt:179]: Do we want to have a permissions test for update storage location here?

File: src/main/kotlin/com/terraformation/backend/customer/api/AdminController.kt (hunk 1 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/customer/api/AdminController.kt:617]: This isn't returning the facility ID; `facility()` is a helper method in this class that returns a redirect to the admin page for the facility. We return a redirect for the usual "don't want to submit the form again on reload" reason.

File: src/main/kotlin/com/terraformation/backend/customer/model/IndividualUser.kt (hunk 2 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/customer/model/IndividualUser.kt:393]: It's all based on previous patterns; the permissions specs don't get down to this level of detail.

File: src/main/kotlin/com/terraformation/backend/customer/model/IndividualUser.kt (hunk 3 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/customer/model/IndividualUser.kt:403]: Good call; I'll do this.

File: src/main/kotlin/com/terraformation/backend/customer/db/FacilityStore.kt (hunk 4 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/customer/db/FacilityStore.kt:193]: Yes, there's a foreign key constraint on `accessions.storage_location_id` that does exactly this.

File: src/main/kotlin/com/terraformation/backend/customer/api/AdminController.kt (hunk 1 above)
Comment (kallen07) [reply src/main/kotlin/com/terraformation/backend/customer/api/AdminController.kt:617]: oh duh! thanks for the response

**Comment by sgrimm** [comment]
**Graphite Merge Job**

Current status: ✅ Merged

This pull request was successfully merged as part of a stack.

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).


Job Reference: DaagJYvBlu4ErBrpiotk

**Pull request merged**